File: Parallel-Java/src/parallel/containers/StackParallel.java
package parallel.containers;

import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.List;

public class StackParallel {
    // template 1, initialize with list
    public ArrayDeque<Character> GetStackWithCharList(List<Character> list){
        ArrayDeque<Character> stack = new ArrayDeque<>(list);
        return stack;
    }

    public ArrayDeque<String> GetStackWithStringList(List<String> list){
        ArrayDeque<String> stack = new ArrayDeque<>(list);
        return stack;
    }

    public ArrayDeque<Integer> GetStackWithIntList(List<Integer> list){
        ArrayDeque<Integer> stack = new ArrayDeque<>(list);
        return stack;
    }

    public ArrayDeque<Double> GetStackWithDoubleList(List<Double> list){
        ArrayDeque<Double> stack = new ArrayDeque<>(list);
        return stack;
    }

    //Template 2, clear if not empty
    public void makeEmptyChar(ArrayDeque<Character> stack){
        if(!stack.isEmpty()){
            stack.clear();
        }
    }

    // Convert Char to String
    public void makeEmptyString(ArrayDeque<String> stack){
        if(!stack.isEmpty()){
            stack.clear();
        }
    }

    // Convert Char to Int
    public void makeEmptyInt(ArrayDeque<Integer> stack){
        if(!stack.isEmpty()){
            stack.clear();
        }
    }

    // Convert Char to Double
    public void makeEmptyDouble(ArrayDeque<Double> stack){
        if(!stack.isEmpty()){
            stack.clear();
        }
    }

    // Template 3, pop until greater than or less than element
    public void findGreaterChar(ArrayDeque<Character> stack, Character element){
        while(!stack.isEmpty()){
            Character c = stack.peek();
            if(c < element){
                stack.pop();
                return;
            }
        }
    }

    public void findLessChar(ArrayDeque<Character> stack, Character element){
        while(!stack.isEmpty()){
            Character c = stack.peek();
            if(c > element){
                stack.pop();
                return;
            }
        }
    }

    public void findEqualChar(ArrayDeque<Character> stack, Character element){
        while(!stack.isEmpty()){
            Character c = stack.peek();
            if(c == element){
                stack.pop();
                return;
            }
        }
    }

    // Convert Char to String
    public void findGreaterString(ArrayDeque<String> stack, String element){
        while(!stack.isEmpty()){
            String c = stack.peek();
            if(c.compareTo(element) < 0){
                stack.pop();
                return;
            }
        }
    }

    public void findLessString(ArrayDeque<String> stack, String element){
        while(!stack.isEmpty()){
            String c = stack.peek();
            if(c.compareTo(element) > 0){
                stack.pop();
                return;
            }
        }
    }

    public void findEqualString(ArrayDeque<String> stack, String element){
        while(!stack.isEmpty()){
            String c = stack.peek();
            if(c.compareTo(element) == 0){
                stack.pop();
                return;
            }
        }
    }

    // Convert Char to Int
    public void findGreaterInt(ArrayDeque<Integer> stack, Integer element){
        while(!stack.isEmpty()){
            Integer c = stack.peek();
            if(c < element){
                stack.pop();
                return;
            }
        }
    }

    public void findLessInt(ArrayDeque<Integer> stack, Integer element){
        while(!stack.isEmpty()){
            Integer c = stack.peek();
            if(c > element){
                stack.pop();
                return;
            }
        }
    }

    public void findEqualInt(ArrayDeque<Integer> stack, Integer element){
        while(!stack.isEmpty()){
            Integer c = stack.peek();
            if(c == element){
                stack.pop();
                return;
            }
        }
    }

    // Convert Char to Double
    public void findGreaterDouble(ArrayDeque<Double> stack, Double element){
        while(!stack.isEmpty()){
            Double c = stack.peek();
            if(c < element){
                stack.pop();
                return;
            }
        }
    }

    public void findLessDouble(ArrayDeque<Double> stack, Double element){
        while(!stack.isEmpty()){
            Double c = stack.peek();
            if(c > element){
                stack.pop();
                return;
            }
        }
    }

    public void findEqualDouble(ArrayDeque<Double> stack, Double element){
        while(!stack.isEmpty()){
            Double c = stack.peek();
            if(c == element){
                stack.pop();
                return;
            }
        }
    }
    
    // Template 4, push list
    public void pushListChar(ArrayDeque<Character> stack, List<Character> list){
        if(list.isEmpty()) return;
        for(Character c : list){
            stack.push(c);
        }
    }

    // Convert Char to String
    public void pushListString(ArrayDeque<String> stack, List<String> list){
        if(list.isEmpty()) return;
        for(String c : list){
            stack.push(c);
        }
    }

    // Convert Char to Int
    public void pushListInt(ArrayDeque<Integer> stack, List<Integer> list){
        if(list.isEmpty()) return;
        for(Integer c : list){
            stack.push(c);
        }
    }
    
    // Convert Char to Double
    public void pushListDouble(ArrayDeque<Double> stack, List<Double> list){
        if(list.isEmpty()) return;
        for(Double c : list){
            stack.push(c);
        }
    }
    
    //Template 5, pop and equal
    public Boolean popEqualChar(ArrayDeque<Character> stack1, ArrayDeque<Character> stack2){
        if(stack1.isEmpty()) return false;
        else stack1.pop();
        Boolean b = Arrays.equals(stack1.toArray(), stack2.toArray());
        return b;
    }

    // Convert String to String
    public Boolean popEqualString(ArrayDeque<String> stack1, ArrayDeque<String> stack2){
        if(stack1.isEmpty()) return false;
        else stack1.pop();
        Boolean b = Arrays.equals(stack1.toArray(), stack2.toArray());
        return b;
    }
    
    // Convert Char to Int
    public Boolean popEqualInt(ArrayDeque<Integer> stack1, ArrayDeque<Integer> stack2){
        if(stack1.isEmpty()) return false;
        else stack1.pop();
        Boolean b = Arrays.equals(stack1.toArray(), stack2.toArray());
        return b;
    }
    
    // Convert Char to Double
    public Boolean popEqualDouble(ArrayDeque<Double> stack1, ArrayDeque<Double> stack2){
        if(stack1.isEmpty()) return false;
        else stack1.pop();
        Boolean b = Arrays.equals(stack1.toArray(), stack2.toArray());
        return b;
    }
}

File: Parallel-Csharp/src/Parallel/Containers/StackParallel.cs
namespace Parallel_Csharp.Parallel.Containers;

public class StackParallel {
    // template 1, initialize with list
    public Stack<char> GetStackWithCharList(List<char> list)
    {
        Stack<char> stack = new Stack<char>(list);
        return stack;
    }
    
    public Stack<string> GetStackWithStringList(List<string> list)
    {
        Stack<string> stack = new Stack<string>(list);
        return stack;
    }
    
    public Stack<int> GetStackWithIntList(List<int> list)
    {
        Stack<int> stack = new Stack<int>(list);
        return stack;
    }
    
    public Stack<double> GetStackWithDoubleList(List<double> list)
    {
        Stack<double> stack = new Stack<double>(list);
        return stack;
    }
    
    //Template 2, clear if not empty
    public void MakeEmptyChar(Stack<char> stack){
        if(stack.Count != 0){
            stack.Clear();
        }
    }

    // Convert Char to String
    public void MakeEmptyString(Stack<string> stack){
        if(stack.Count != 0){
            stack.Clear();
        }
    }
    
    // Convert Char to Int
    public void MakeEmptyInt(Stack<int> stack){
        if(stack.Count != 0){
            stack.Clear();
        }
    }
    
    // Convert Char to Double
    public void MakeEmptyDouble(Stack<double> stack){
        if(stack.Count != 0){
            stack.Clear();
        }
    }
    
    // Template 3, pop until greater than or less than element
    public void FindGreaterChar(Stack<char> stack, char element){
        while(stack.Count != 0){
            char c = stack.Peek();
            if(c < element){
                stack.Pop();
                return;
            }
        }
    }

    public void FindLessChar(Stack<char> stack, char element){
        while(stack.Count != 0){
            char c = stack.Peek();
            if(c > element){
                stack.Pop();
                return;
            }
        }
    }
    
    public void FindEqualChar(Stack<char> stack, char element){
        while(stack.Count == 0){
            char c = stack.Peek();
            if(c > element){
                stack.Pop();
                return;
            }
        }
    }

    // todo Convert Char to String
    public void FindGreaterString(Stack<String> stack, String element){
        while(stack.Count != 0){
            String c = stack.Peek();
            if(c.CompareTo(element) < 0){
                stack.Pop();
                return;
            }
        }
    }

    public void FindLessString(Stack<String> stack, String element){
        while(stack.Count != 0){
            String c = stack.Peek();
            if(c.CompareTo(element) > 0){
                stack.Pop();
                return;
            }
        }
    }
    
    public void FindEqualString(Stack<String> stack, String element){
        while(stack.Count == 0){
            String c = stack.Peek();
            if(c.CompareTo(element) == 0){
                stack.Pop();
                return;
            }
        }
    }

    // todo Convert Char to Int
    public void FindGreaterInt(Stack<int> stack, int element){
        while(stack.Count != 0){
            int c = stack.Peek();
            if(c < element){
                stack.Pop();
                return;
            }
        }
    }

    public void FindLessInt(Stack<int> stack, int element){
        while(stack.Count != 0){
            int c = stack.Peek();
            if(c > element){
                stack.Pop();
                return;
            }
        }
    }
    
    public void FindEqualInt(Stack<int> stack, int element){
        while(stack.Count == 0){
            int c = stack.Peek();
            if(c > element){
                stack.Pop();
                return;
            }
        }
    }

    // todo Convert Char to Double
    public void FindGreaterDouble(Stack<double> stack, double element){
        while(stack.Count != 0){
            double c = stack.Peek();
            if(c < element){
                stack.Pop();
                return;
            }
        }
    }

    public void FindLessDouble(Stack<double> stack, double element){
        while(stack.Count != 0){
            double c = stack.Peek();
            if(c > element){
                stack.Pop();
                return;
            }
        }
    }
    
    public void FindEqualDouble(Stack<double> stack, double element){
        while(stack.Count == 0){
            double c = stack.Peek();
            if(c > element){
                stack.Pop();
                return;
            }
        }
    }

    // Template 4, push list
    public void PushListChar(Stack<char> stack, List<char> list){
        if(list.Count == 0) return;
        foreach(char c in list){
            stack.Push(c);
        }
    }

    // Convert Char to String
    public void PushListString(Stack<string> stack, List<string> list){
        if(list.Count == 0) return;
        foreach(string c in list){
            stack.Push(c);
        }
    }
    
    // Convert Char to Int
    public void PushListInt(Stack<int> stack, List<int> list){
        if(list.Count == 0) return;
        foreach(int c in list){
            stack.Push(c);
        }
    }
    
    // Convert Char to Double
    public void PushListDouble(Stack<double> stack, List<double> list){
        if(list.Count == 0) return;
        foreach(double c in list){
            stack.Push(c);
        }
    }
    
    //Template 5, pop and equal
    public bool PopEqualChar(Stack<char> stack1, Stack<char> stack2){
        if(stack1.Count == 0) return false;
        else stack1.Pop();
        bool b = stack1.SequenceEqual(stack2);
        return b;
    }

    // Convert Char to String
    public bool PopEqualString(Stack<string> stack1, Stack<string> stack2){
        if(stack1.Count == 0) return false;
        else stack1.Pop();
        bool b = stack1.SequenceEqual(stack2);
        return b;
    }
    
    // Convert Char to Int
    public bool PopEqualInt(Stack<int> stack1, Stack<int> stack2){
        if(stack1.Count == 0) return false;
        else stack1.Pop();
        bool b = stack1.SequenceEqual(stack2);
        return b;
    }
    
    // Convert Char to Double
    public bool PopEqualDouble(Stack<double> stack1, Stack<double> stack2){
        if(stack1.Count == 0) return false;
        else stack1.Pop();
        bool b = stack1.SequenceEqual(stack2);
        return b;
    }
}